### DISCUSSION TITLE
[Feature] Losslessly and reversibly convert existing and new jpeg to jxl

### DISCUSSION THREAD

**Discussion by Qhilm** [immich-app/immich#12266] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I know this goes against the "never touch the original" mantra, but this saves a whooping 25% storage space on average and is a _lossless_ and moreover **reversible** operation (must use the `--lossless_jpeg=1` option). Of course it should be opt-in.

There are two parts to this:
1. for new files: upon uploading, if the type is jpeg, losslessly convert it and store it as jxl.
2. For existing file, have a utility added to the "utilities" section called "Losslessly convert JPEG to JPEG XL".

Ideally, the level of effort should be configurable.

Command sample:

`cjxl IMG123.jpeg IMG123.jxl --distance=0 --lossless_jpeg=1 --effort=7`

It would maybe be worth putting a note (a tooltip ⓘ for example) that any effort level above 7 only bring marginal benefits in terms of storage while causing ten or hundredfold CPU load.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by radh21301** [comment]
I am not technically adept to comment of the technicalities of the lossless jpeg. Though it sounds pretty awesome with the way you describe it. 


My take is more of a philosophical one.
I think what works very much to Immich's advantage, is a blanket statement that "We NEVER touch your originals" - just like you said. I think, the main point is that it is a blanket statement - will apply to absolutely everything in Immich. Once you have features that do not follow this, even if they are opt in - would move Immich away from this blanket statement. Not that it is a terrible thing, but it puts immich in a grey area.

For example: When someone who wants to start with Immich asks a question "Hey guys, will immich do something to my images - edit, process, etc", the Immich dev's answer would be "We dont do anything to your original, unless you enables features X,Y,Z which can edit your images. They are not enabled by default, so dont worry". Then that person would probably be in a spot where they need to research all the features that could possibly edit images, and ensure that they do are not enabled. This would also need to apply to all new features that Immich comes up with in the future.

I am sure the dev's would take care to ensure this is always OPT-IN. Even then, the huge advantage that Immich has 'we never touch originals' will be lost.

Sorry for the ramble. 


Maybe, this could be a plugin that integrates with Immich. The devs have something like this envisioned where people can develop plugins that works seamlessly with immich.

**Reply by Qhilm** [reply]
I see your point, but the answer to the question you mentioned is already "unless" to a certain extent:
* Unless you mess up the storage template
* Unless you delete the picture by mistake

and I'm sure there's more.

**Reply by radh21301** [reply]
'Delete' option deleting images is  an expected behavior.

But messing up storage templates will not delete your images. How can it? All it does is move images around right? If it does delete, it is a bug which will be fixed.

I would not put these in the same category as converting an image to a different format.

**Comment by bo0tzz** [comment]
#3996
